### ISSUE TITLE
Ignored field prefix in TF JSON?

### ISSUE THREAD

**Issue by eropple** [hashicorp/terraform#784]
When generating JSON for Terraform, it'd be helpful to be able to insert fields that TF can ignore--debugging information, so on and so forth. Example:

``` json
{
  "//GEN": "Built by Terraframe 0.1.0",
  "resource": {
    "aws_instance": {
      "infra_test_vpn": {
        "//SOURCE": "vpn.terraframe:4",
        "tags": {
          "Name": "infra_test_vpn"
        },
        "ami": "ami-76e27e1e",
        "availability_zone": "us-east-1b",
        "instance_type": "t2.micro",
        "key_name": "eropple",
        "security_groups": [
          "${aws_security_group.infra_test_vpn.id}"
        ],
        "subnet_id": "${aws_subnet.us-east-1b_infra_test_public.id}",
        "provisioner": {
          "file": [
            {
              "source": "/tmp/testfile",
              "destination": "/tmp/landingtest",
              "connection": {
                "user": "ubuntu",
                "key_file": "/Users/Ed/.ssh/id_rsa"
              }
            }
          ]
        }
      }
    }
  }
}
```

This JSON actually fails in Terraform 0.3.6 right now - different issue! - but being able to backtrack through such failures in a comprehensible manner would be really helpful.

**Comment by mitchellh** [comment]
I think this is a good idea that we should just adopt into Terraform itself. I think the `//` syntax is a bit clunky within the JSON, but maybe thats me being unreasonable and bikeshedding this. 

What about any key named `__comment` is stripped out? Since its not invalid JSON to repeat a key multiple times even within the same object, this would probably work.

**Comment by eropple** [comment]
I forgot repeated keys were legal. Works for me!

**Comment by mitchellh** [comment]
This is turning out to be more annoying than I anticipated. I'll try again in the future, but this will help anyone (future me perhaps) who is trying to do this in the future:

```go
// This will delete any comments from the configuration (fields matching
// "__comment"). This modifies the parameter directly.
func hclDeleteComments(m map[string]interface{}) map[string]interface{} {
	walker := &hclDeleteKeyWalker{Key: HCLJSONCommentKey}
	if err := reflectwalk.Walk(m, walker); err != nil {
		panic(err)
	}

	return m
}

type hclDeleteKeyWalker struct {
	Key string
}

func (w *hclDeleteKeyWalker) Map(m reflect.Value) error {
	return nil
}

func (w *hclDeleteKeyWalker) MapElem(m, k, v reflect.Value) error {
	if k.Kind() != reflect.String {
		return nil
	}

	if k.String() == w.Key {
		var zero reflect.Value
		m.SetMapIndex(k, zero)
	}

	return nil
}
```

**Comment by glenjamin** [comment]
I just ran into this, and came up with this workaround:

```json
{
  "variable": {
    "users.tf.rb": {
      "default": "Generated from ruby source by terrafiddle"
    }
  },
}
```

If I understand correctly, the proposal would be to bake support for ignoring keys into the `hcl`project? I might have a crack at this.

**Comment by apparentlymart** [comment]
[The next major release](https://www.hashicorp.com/blog/terraform-0-1-2-preview) will include a new JSON engine that, among other things, accepts a special property name `"//"` in any object that's representing a block body or root body (not objects representing maps) but then entirely ignores it, allowing it to be used as a comment.

```json
{
  "//": "Generated by foobarbaz",
  "variable": {
    "example": {}
  }
}
```

**Comment by apparentlymart** [comment]
I verified this in Terraform v0.12-alpha1 using the following example configuration:

```json
{
    "//": "Generated by example.baz; DO NOT EDIT!",
    "provider": {
        "aws": {
            "//": "The main AWS provider is for us-east-1 because that's where our main admin tools live",
            "region": "us-east-1",

            "//": "All actions in this region must be done with the admin role",
            "assume_role": "arn:blah:blah:admin"
        }
    },
    "resource": {
        "aws_instance": {
            "foo": {
                "//": "This is our main compute instance",
                "ami": "ami-12353",
                "instance_type": "t2.micro"
            }
        }
    }
}
```

As I mentioned in my previous comment above, the special property name `"//"` is ignored by the configuration language parser in contexts that would be nested blocks in the native syntax.

This feature will also be included in the final release of v0.12.0.
